citro3d/src/uniform.rs

```diff
@@ -20,38 +22,147 @@ impl From<Index> for i32 {
     }
 }
 
-mod private {
-    use crate::math::Matrix;
+/// A uniform which may be bound as input to a shader program
+pub enum Uniform {
+    /// Single float uniform (`.fvec name`)
+    Float(FVec4),
+    /// Two element float uniform (`.fvec name[2]`)
+    Float2([FVec4; 2]),
+    /// Three element float uniform (`.fvec name [3]`)
+    Float3([FVec4; 3]),
+    /// Matrix/4 element float uniform (`.fvec name[4]`)
+    Float4(Matrix4),
+    /// Bool uniform (`.bool name`)
+    Bool(bool),
+    /// Integer uniform (`.ivec name`)
+    Int(IVec),
+}
+impl Uniform {
+    /// Get range of valid indexes for this uniform to bind to
+    pub fn index_range(&self) -> Range<Index> {
+        // these indexes are from the uniform table in the shader see: https://www.3dbrew.org/wiki/SHBIN#Uniform_Table_Entry
+        // the input registers then are excluded by libctru, see: https://github.com/devkitPro/libctru/blob/0da8705527f03b4b08ff7fee4dd1b7f28df37905/libctru/source/gpu/shbin.c#L93
+        match self {
+            Uniform::Float(_) | Uniform::Float2(_) | Uniform::Float3(_) | Uniform::Float4(_) => {
+                Index(0)..Index(0x60)
+            }
+            Uniform::Int(_) => Index(0x60)..Index(0x64),
+            // this gap is intentional
+            Uniform::Bool(_) => Index(0x68)..Index(0x79),
+        }
+    }
+    /// Get length of uniform, i.e. how many registers it will write to
+    #[allow(clippy::len_without_is_empty)] // is_empty doesn't make sense here
+    pub fn len(&self) -> usize {
+        match self {
+            Uniform::Float(_) => 1,
+            Uniform::Float2(_) => 2,
+            Uniform::Float3(_) => 3,
+            Uniform::Float4(_) => 4,
+            Uniform::Bool(_) | Uniform::Int(_) => 1,
+        }
+    }
 
-    pub trait Sealed {}
+    /// Bind a uniform
+    ///
+    /// Note: `_instance` is here to ensure unique access to the global uniform buffers
+    /// otherwise we could race and/or violate aliasing
+    pub(crate) fn bind(self, _instance: &mut Instance, ty: shader::Type, index: Index) {
+        assert!(
+            self.index_range().contains(&index),
+            "tried to bind uniform to an invalid index (index: {}, valid range: {}..{})",
```

Comment:
nit: I think you can just use `{:?}` with `self.index_range()` to get the default Range Debug impl

Reply:
yeah I don't think `Index` has a debug impl? I really should just add a `Display` impl for it tbh and use that I think I kept avoiding doing that for some reason